### PR TITLE
Redox OS support

### PR DESCRIPTION
**Pull request by jD91mZM2**
Might be worth releasing a libc update first before to avoid using the git version

### CODE REVIEW

**Comment by alexcrichton** [comment]
Since there's a lot of `#[cfg]` going on here, can redox be moved to a separate file like windows?

**Comment by jD91mZM2** [comment]
That's probably a good idea, although I'm a little worried it'll be a lot of repeated code anyway. The unix file is 1000 lines long and eventually redox will gain support for most of this I assume.

**Comment by jD91mZM2** [comment]
Alright, there we go. It would be cool if I could somehow reuse the functions that are the same on both, but I assume rust wouldn't like `use`ing the unix file on redox.

**Comment by jD91mZM2** [comment]
@alexcrichton Bump?

File: src/socket.rs [hunk 1]
```diff
@@ -827,23 +827,27 @@ impl Type {
         Type(c::SOCK_DGRAM)
     }
 
+    #[cfg(not(target_os = "redox"))]
```
Comment (alexcrichton) [inline src/socket.rs]: This is a public API and so I'd prefer to not change the public interface of this crate depending on the platform currently. Could the redox module use a sentinel value for this constant which causes a runtime error when used in this crate?

Comment (jD91mZM2) [reply src/socket.rs]: Good point! Updated :)

**Comment by alexcrichton** [comment]
Er sorry so my goal here is to reduce the amount of `#[cfg]` here as much as possible because it otherwise adds maintenance burden to the rest of the code on other platforms. Instead of panicking could a runtime error be issued only when a socket is attempted to be created with an invalid protocol not supported?

**Comment by alexcrichton** [comment]
Thanks!

**Comment by jD91mZM2** [comment]
Awesome! Would be great if you could bump the version of this & libc. Sorry for being annoying.

**Comment by alexcrichton** [comment]
I've published these changes and the libc changes are happening in the libc repo

**Pull request merged**